### PR TITLE
Automatically log in after registration

### PR DESCRIPTION
**Pull request by justjanne**
Type: Defect
Closes: https://github.com/vector-im/element-web/issues/19305

## In Short

* Stop using `inhibit_login`, instead make the decision which client gets to log in *after* we've received the tokens
* Update the UI accordingly

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Automatically log in after registration ([\#8654](https://github.com/matrix-org/matrix-react-sdk/pull/8654)). Fixes vector-im/element-web#19305. Contributed by @justjanne.

### CODE REVIEW

File: src/components/structures/auth/Registration.tsx [hunk 1]
```diff
@@ -39,6 +39,14 @@ import Spinner from "../../views/elements/Spinner";
 import { AuthHeaderDisplay } from './header/AuthHeaderDisplay';
 import { AuthHeaderProvider } from './header/AuthHeaderProvider';
 
+const DEBUG = false;
+let debuglog = function(msg: string, ...args: any[]) {};
+
+if (DEBUG) {
+    // using bind means that we get to keep useful line numbers in the console
+    debuglog = logger.log.bind(console);
+}
```
Comment (ShadowJonathan) [inline src/components/structures/auth/Registration.tsx:48]: Why is this necessary? Can't `console.debug` be used?

File: src/components/structures/auth/Registration.tsx [hunk 2]
```diff
@@ -288,6 +296,7 @@ export default class Registration extends React.Component<IProps, IState> {
     };
 
     private onUIAuthFinished = async (success: boolean, response: any) => {
```
Comment (t3chguy) [inline src/components/structures/auth/Registration.tsx:298]: Can we fix this `any`? It seems like it caused a bug prior to this change where it was checking `userId`

**Comment by t3chguy** [comment]
Do you think it is feasible to wire up a Cypress test for this? Could use something like https://hub.docker.com/r/mailhog/mailhog/

File: src/components/structures/auth/Registration.tsx (hunk 2 above)
Comment (justjanne) [reply src/components/structures/auth/Registration.tsx:298]: Sure, I’ll try to figure out the real type and update it :)

**Comment by justjanne** [comment]
> Do you think it is feasible to wire up a Cypress test for this?

I can try, but I can’t promise results^^

**Comment by t3chguy** [comment]
image

This doesn't feel appropriate given I was just auto-logged in on the other window, maybe it should be more of a "You can close this now and enjoy Element in the other tab/window"

File: src/components/structures/auth/Registration.tsx (hunk 1 above)
Comment (ShadowJonathan) [inline src/components/structures/auth/Registration.tsx:48]: Any rationalization for why this bit is needed?

Comment (t3chguy) [reply src/components/structures/auth/Registration.tsx:48]: Also, it created 3 code smells in Sonar

**Comment by justjanne** [comment]
@t3chguy If you click "log in", you get automatically logged in in that window as well. That was also part of the flow previously, I could’ve just made it automatically redirect you to the home page, but I wasn’t sure whether that was appropriate

File: src/components/structures/auth/Registration.tsx (hunk 1 above)
Comment (justjanne) [reply src/components/structures/auth/Registration.tsx:48]: It’s our standard solution for debug logging across the entire codebase, it’s used in countless files. I didn’t invent this, but while I don’t specifically like it, I didn’t want to experiment with debug logging as well (this ticket has enough things it could negatively affect already)

File: src/components/structures/auth/Registration.tsx (hunk 2 above)
Comment (justjanne) [reply src/components/structures/auth/Registration.tsx:298]: Yeah, doesn’t look like it’ll really be possible to do this properly.

This entire part of the JS SDK is still missing types and the types used here aren’t even close to what they should be (just one example: https://github.com/matrix-org/matrix-js-sdk/blob/develop/src/client.ts#L6840-L6849)

File: src/components/structures/auth/Registration.tsx (hunk 1 above)
Comment (t3chguy) [reply src/components/structures/auth/Registration.tsx:48]: Ftr its the older standard, not much new code uses it
One new approach is
https://github.com/matrix-org/matrix-react-sdk/blob/46ba14219c7d7f40120784bb8324106a1c44cc35/src/components/structures/ScrollPanel.tsx#L38-L42

File: src/components/structures/auth/Registration.tsx (hunk 2 above)
Comment (t3chguy) [reply src/components/structures/auth/Registration.tsx:297]: Can we at least type the fields we use in this method and label it as partial? The `room_id` vs `roomId` is pretty disastrous.

Comment (justjanne) [reply src/components/structures/auth/Registration.tsx:297]: I could, but that’d still first require a change in matrix-js-sdk (we’d need to add the required fields to the AuthData interface), which would definitely be a T-Enhancement, so it’d require PM approval.

I’ve created branches with the minimal changes that’d be required for this:

https://github.com/matrix-org/matrix-js-sdk/commit/d7fa76a34afb796ebce97301b9ee4d5c8f3cc9ac
https://github.com/matrix-org/matrix-react-sdk/commit/3c2e1f5d1b0926454953209b601d77d8810da8db

Comment (t3chguy) [reply src/components/structures/auth/Registration.tsx:297]: Adding types is `T-Task` - not `T-Enhancement` and thus does not require anything beyond code review.

File: src/components/views/dialogs/DeactivateAccountDialog.tsx [hunk 3]
```diff
@@ -104,7 +104,7 @@ export default class DeactivateAccountDialog extends React.Component<IProps, ISt
         this.setState({ bodyText, continueText, continueKind });
     };
 
-    private onUIAuthFinished = (success: boolean, result: Error) => {
+    private onUIAuthFinished = (success, result) => {
```
Comment (t3chguy) [inline src/components/views/dialogs/DeactivateAccountDialog.tsx:107]: Why did we drop the types here?

File: src/components/views/dialogs/InteractiveAuthDialog.tsx [hunk 4]
```diff
@@ -117,7 +117,7 @@ export default class InteractiveAuthDialog extends React.Component<IProps, IStat
         };
     }
 
-    private onAuthFinished = (success: boolean, result: Error): void => {
+    private onAuthFinished = (success, result): void => {
```
Comment (t3chguy) [inline src/components/views/dialogs/InteractiveAuthDialog.tsx:120]: Same as above

**Review by t3chguy — APPROVED** [review]
Looks sane otherwise

File: src/components/views/dialogs/DeactivateAccountDialog.tsx (hunk 3 above)
Comment (justjanne) [reply src/components/views/dialogs/DeactivateAccountDialog.tsx]: Because I was switching to `private onUIAuthFinished: InteractiveAuthCallback = (success, result) => {` but forgot to actually finish that part.
